Gradle Dependency Catalog compatibility

Hello 

I enabled Gradle Dependency Catalog with `1.5.0-RC-2` and I got following stacktrace.  

```
FAILURE: Build failed with an exception.

* What went wrong:
Using dependency catalogs requires the activation of the matching feature preview.
See the documentation at https://docs.gradle.org/7.2/userguide/platforms.html#sub:central-declaration-of-dependencies

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 560ms
```

I also see it deleted my projects root `build.gradle` and `settings.gradle`

`git status`

```
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	deleted:    build.gradle
	deleted:    settings.gradle

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	backup_build.gradle
	backup_settings.gradle

no changes added to commit (use "git add" and/or "git commit -a")
```

Inside the `settings.gradle` it contains the line to active the Gradle Dependency Catalog

```
enableFeaturePreview("VERSION_CATALOGS")
enableFeaturePreview("TYPESAFE_PROJECT_ACCESSORS")
```

Any idea?

I'm using Gradle 7.2 btw

Thanks in advance

I followed the installation steps on the Github page so I only have a `mirakle.gradle` file in my `$HOME/.gradle/init.d/mirakle_init.gradle`. Nothing inside my project root

The contents of the `$HOME/.gradle/init.d/mirakle_init.gradle` file

```
initscript {
    repositories {
        mavenCentral()
    }
    dependencies {
        classpath 'io.github.adambl4:mirakle:1.5.0-RC-2'
    }
}

apply plugin: Mirakle

rootProject {

    mirakle {
        host "macmini"
        rsyncToRemoteArgs += ["-avzhP", "--no-p", "--exclude='.git'", "--exclude='.gradle'"]
        rsyncFromRemoteArgs += ["-avzhP", "--no-p", "--exclude='.git'", "--exclude='.gradle'"]
        rsyncFromRemoteArgs += ["-q", "-i", "-v", "--progress"]
        downloadInParallel true
        downloadInterval 2000
    }
}
```

I start following task on the command line:

`./gradlew clean --stacktrace`

Which results in following stacktrace.

```                                                                                                      
FAILURE: Build failed with an exception.

* What went wrong:
Using dependency catalogs requires the activation of the matching feature preview.
See the documentation at https://docs.gradle.org/7.2/userguide/platforms.html#sub:central-declaration-of-dependencies

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.InvalidUserCodeException: Using dependency catalogs requires the activation of the matching feature preview.
See the documentation at https://docs.gradle.org/7.2/userguide/platforms.html#sub:central-declaration-of-dependencies
        at org.gradle.internal.management.DefaultVersionCatalogBuilderContainer.create(DefaultVersionCatalogBuilderContainer.java:85)
        at org.gradle.internal.management.DefaultVersionCatalogBuilderContainer.create(DefaultVersionCatalogBuilderContainer.java:46)
        at org.gradle.api.internal.AbstractNamedDomainObjectContainer.create(AbstractNamedDomainObjectContainer.java:56)
        at org.gradle.configuration.BuildTreePreparingProjectsPreparer.lambda$generateDependenciesAccessorsAndAssignPluginVersions$0(BuildTreePreparingProjectsPreparer.java:87)
        at org.gradle.internal.management.DefaultDependencyResolutionManagement.versionCatalogs(DefaultDependencyResolutionManagement.java:134)
        at org.gradle.configuration.BuildTreePreparingProjectsPreparer.generateDependenciesAccessorsAndAssignPluginVersions(BuildTreePreparingProjectsPreparer.java:84)
        at org.gradle.configuration.BuildTreePreparingProjectsPreparer.prepareProjects(BuildTreePreparingProjectsPreparer.java:54)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer$ConfigureBuild.run(BuildOperationFiringProjectsPreparer.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:56)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$run$1(DefaultBuildOperationExecutor.java:74)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.runWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:45)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:74)
        at org.gradle.configuration.BuildOperationFiringProjectsPreparer.prepareProjects(BuildOperationFiringProjectsPreparer.java:40)
        at org.gradle.initialization.VintageBuildModelController.prepareProjects(VintageBuildModelController.java:97)
        at org.gradle.initialization.VintageBuildModelController.doBuildStages(VintageBuildModelController.java:77)
        at org.gradle.initialization.VintageBuildModelController.prepareToScheduleTasks(VintageBuildModelController.java:64)
        at org.gradle.internal.build.DefaultBuildLifecycleController.lambda$prepareToScheduleTasks$0(DefaultBuildLifecycleController.java:104)
        at org.gradle.internal.build.DefaultBuildLifecycleController.lambda$withModelOrThrow$5(DefaultBuildLifecycleController.java:142)
        at org.gradle.internal.build.DefaultBuildLifecycleController.withModel(DefaultBuildLifecycleController.java:157)
        at org.gradle.internal.build.DefaultBuildLifecycleController.withModelOrThrow(DefaultBuildLifecycleController.java:140)
        at org.gradle.internal.build.DefaultBuildLifecycleController.prepareToScheduleTasks(DefaultBuildLifecycleController.java:102)
        at org.gradle.internal.buildtree.DefaultBuildTreeWorkPreparer.scheduleRequestedTasks(DefaultBuildTreeWorkPreparer.java:33)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.lambda$doScheduleAndRunTasks$1(DefaultBuildTreeLifecycleController.java:83)
        at org.gradle.composite.internal.DefaultIncludedBuildTaskGraph.withNewTaskGraph(DefaultIncludedBuildTaskGraph.java:94)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.doScheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:82)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.runBuild(DefaultBuildTreeLifecycleController.java:104)
        at org.gradle.internal.buildtree.DefaultBuildTreeLifecycleController.scheduleAndRunTasks(DefaultBuildTreeLifecycleController.java:64)
        at org.gradle.tooling.internal.provider.ExecuteBuildActionRunner.run(ExecuteBuildActionRunner.java:31)
        at org.gradle.launcher.exec.ChainingBuildActionRunner.run(ChainingBuildActionRunner.java:35)
        at org.gradle.internal.buildtree.ProblemReportingBuildActionRunner.run(ProblemReportingBuildActionRunner.java:47)
        at org.gradle.launcher.exec.BuildOutcomeReportingBuildActionRunner.run(BuildOutcomeReportingBuildActionRunner.java:69)
        at org.gradle.tooling.internal.provider.FileSystemWatchingBuildActionRunner.run(FileSystemWatchingBuildActionRunner.java:91)
        at org.gradle.launcher.exec.BuildCompletionNotifyingBuildActionRunner.run(BuildCompletionNotifyingBuildActionRunner.java:41)
        at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.lambda$execute$0(RootBuildLifecycleBuildActionExecutor.java:40)
        at org.gradle.composite.internal.DefaultRootBuildState.run(DefaultRootBuildState.java:154)
        at org.gradle.launcher.exec.RootBuildLifecycleBuildActionExecutor.execute(RootBuildLifecycleBuildActionExecutor.java:40)
        at org.gradle.internal.buildtree.DefaultBuildTreeContext.execute(DefaultBuildTreeContext.java:40)
        at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.lambda$execute$0(BuildTreeLifecycleBuildActionExecutor.java:56)
        at org.gradle.internal.buildtree.BuildTreeState.run(BuildTreeState.java:53)
        at org.gradle.launcher.exec.BuildTreeLifecycleBuildActionExecutor.execute(BuildTreeLifecycleBuildActionExecutor.java:56)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:61)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor$3.call(RunAsBuildOperationBuildActionExecutor.java:57)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:200)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:195)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:75)
        at org.gradle.internal.operations.DefaultBuildOperationRunner$3.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:153)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:68)
        at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:62)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.lambda$call$2(DefaultBuildOperationExecutor.java:79)
        at org.gradle.internal.operations.UnmanagedBuildOperationWrapper.callWithUnmanagedSupport(UnmanagedBuildOperationWrapper.java:54)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:79)
        at org.gradle.launcher.exec.RunAsBuildOperationBuildActionExecutor.execute(RunAsBuildOperationBuildActionExecutor.java:57)
        at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.lambda$execute$0(RunAsWorkerThreadBuildActionExecutor.java:38)
        at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:211)
        at org.gradle.launcher.exec.RunAsWorkerThreadBuildActionExecutor.execute(RunAsWorkerThreadBuildActionExecutor.java:38)
        at org.gradle.tooling.internal.provider.ContinuousBuildActionExecutor.execute(ContinuousBuildActionExecutor.java:103)
        at org.gradle.tooling.internal.provider.SubscribableBuildActionExecutor.execute(SubscribableBuildActionExecutor.java:64)
        at org.gradle.internal.session.DefaultBuildSessionContext.execute(DefaultBuildSessionContext.java:46)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter.lambda$execute$0(BuildSessionLifecycleBuildActionExecuter.java:56)
        at org.gradle.internal.session.BuildSessionState.run(BuildSessionState.java:69)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter.execute(BuildSessionLifecycleBuildActionExecuter.java:55)
        at org.gradle.tooling.internal.provider.BuildSessionLifecycleBuildActionExecuter.execute(BuildSessionLifecycleBuildActionExecuter.java:37)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:36)
        at org.gradle.tooling.internal.provider.GradleThreadBuildActionExecuter.execute(GradleThreadBuildActionExecuter.java:25)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:63)
        at org.gradle.tooling.internal.provider.StartParamsValidatingActionExecuter.execute(StartParamsValidatingActionExecuter.java:31)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:58)
        at org.gradle.tooling.internal.provider.SessionFailureReportingActionExecuter.execute(SessionFailureReportingActionExecuter.java:42)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:47)
        at org.gradle.tooling.internal.provider.SetupLoggingActionExecuter.execute(SetupLoggingActionExecuter.java:31)
        at org.gradle.launcher.daemon.server.exec.ExecuteBuild.doBuild(ExecuteBuild.java:65)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.WatchForDisconnection.execute(WatchForDisconnection.java:39)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ResetDeprecationLogger.execute(ResetDeprecationLogger.java:29)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.RequestStopIfSingleUsedDaemon.execute(RequestStopIfSingleUsedDaemon.java:35)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:78)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput$2.create(ForwardClientInput.java:75)
        at org.gradle.util.internal.Swapper.swap(Swapper.java:38)
        at org.gradle.launcher.daemon.server.exec.ForwardClientInput.execute(ForwardClientInput.java:75)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogAndCheckHealth.execute(LogAndCheckHealth.java:55)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.LogToClient.doBuild(LogToClient.java:63)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.EstablishBuildEnvironment.doBuild(EstablishBuildEnvironment.java:84)
        at org.gradle.launcher.daemon.server.exec.BuildCommandOnly.execute(BuildCommandOnly.java:37)
        at org.gradle.launcher.daemon.server.api.DaemonCommandExecution.proceed(DaemonCommandExecution.java:104)
        at org.gradle.launcher.daemon.server.exec.StartBuildOrRespondWithBusy$1.run(StartBuildOrRespondWithBusy.java:52)
        at org.gradle.launcher.daemon.server.DaemonStateCoordinator$1.run(DaemonStateCoordinator.java:297)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:61)


* Get more help at https://help.gradle.org

BUILD FAILED in 540ms
```

Inside my project's `settings.gradle` I have enabled following incubating Gradle features

```
enableFeaturePreview("VERSION_CATALOGS")
enableFeaturePreview("TYPESAFE_PROJECT_ACCESSORS")
```

I'll try with an empty project and enable Gradle Version Catalog to see if it is project settings related.

--- a/benchmark/build.gradle
+++ b/benchmark/build.gradle
@@ -11,9 +11,8 @@ repositories {
 }
 
 dependencies {
-    compile gradleApi()
-    compile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
-    compile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    implementation gradleApi()
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
 }
 
 sourceSets {

--- a/build.gradle
+++ b/build.gradle
@@ -1,8 +1,8 @@
 group 'com.instamotor'
-version '1.4.3'
+version '1.5.0-RC-4'
 
 buildscript {
-    ext.kotlin_version = '1.4.32'
+    ext.kotlin_version = '1.5.31'
 
     repositories {
         mavenCentral()

--- a/plugin-test/build.gradle
+++ b/plugin-test/build.gradle
@@ -24,12 +24,12 @@ repositories {
 }
 
 dependencies {
-    compile project(":mirakle")
-    testCompile gradleTestKit()
-    testCompile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
-    testCompile "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
-    testCompile "org.jetbrains.kotlin:kotlin-test:$kotlin_version"
-    testCompile 'org.jetbrains.spek:spek-api:1.1.5'
-    testRuntime "org.junit.platform:junit-platform-launcher:1.0.0"
-    testRuntime 'org.jetbrains.spek:spek-junit-platform-engine:1.1.5'
+    implementation project(":mirakle")
+    testImplementation gradleTestKit()
+    testImplementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    testImplementation "org.jetbrains.kotlin:kotlin-reflect:$kotlin_version"
+    testImplementation "org.jetbrains.kotlin:kotlin-test:$kotlin_version"
+    testImplementation 'org.jetbrains.spek:spek-api:1.1.5'
+    //testRuntime "org.junit.platform:junit-platform-launcher:1.0.0"
+    //testRuntime 'org.jetbrains.spek:spek-junit-platform-engine:1.1.5'
 }

--- a/plugin/build.gradle
+++ b/plugin/build.gradle
@@ -5,9 +5,14 @@ compileKotlin {
     kotlinOptions {
         useIR = true
         jvmTarget = "1.8"
-        languageVersion = "1.4"
+        languageVersion = "1.5"
         freeCompilerArgs += ["-Xno-optimized-callable-references",]
     }
+}
+
+tasks.withType(JavaCompile).configureEach {
+    sourceCompatibility = JavaVersion.VERSION_1_8
+    targetCompatibility = JavaVersion.VERSION_1_8
 }
 
 repositories {
@@ -15,10 +20,10 @@ repositories {
 }
 
 dependencies {
-    compile gradleApi()
-    compile "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
-    compile 'com.googlecode.streamflyer:streamflyer-core:1.1.3'
-    compile 'com.googlecode.streamflyer-regex-fast:streamflyer-regex-fast:1.0.1'
+    implementation gradleApi()
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:$kotlin_version"
+    implementation 'com.googlecode.streamflyer:streamflyer-core:1.1.3'
+    implementation 'com.googlecode.streamflyer-regex-fast:streamflyer-regex-fast:1.0.1'
 }
 
 sourceSets {

--- a/plugin/src/main/kotlin/Mirakle.kt
+++ b/plugin/src/main/kotlin/Mirakle.kt
@@ -62,26 +62,38 @@ open class Mirakle : Plugin<Gradle> {
             }
 
             if (!breakMode) {
-                //a way to make Gradle not evaluate build.gradle and settings.gradle on local machine
+                //a way to make Gradle not evaluate build.gradle, settings.gradle and dependency catalog files on local machine
 
-                val settingsFile = gradlewRoot.listFiles().firstOrNull { it.name.startsWith("settings.gradle") }
+                val settingsFile = gradlewRoot.listFiles()?.firstOrNull { it.name.startsWith("settings.gradle") }
                 val settingsBackup = settingsFile?.let { File(gradlewRoot, "backup_${it.name}") }
 
-                val buildFile = gradlewRoot.listFiles().firstOrNull { it.name.startsWith("build.gradle") }
+                val buildFile = gradlewRoot.listFiles()?.firstOrNull { it.name.startsWith("build.gradle") }
                 val buildFileBackup = buildFile?.let { File(gradlewRoot, "backup_${it.name}") }
+
+                val versionCatalog = File(gradlewRoot, "gradle").listFiles()?.filter {
+                    it.name.endsWith("versions.toml")
+                }
 
                 settingsFile?.renameTo(settingsBackup)
                 buildFile?.renameTo(buildFileBackup)
 
-                gradle.rootProject {
-                    it.afterEvaluate {
-                        settingsFile?.let {
-                            settingsBackup?.renameTo(it)
-                        }
+                val versionCatalogBackup = versionCatalog?.map {
+                    val backup = File(it.parent, "${it.name}_backup")
+                    it.renameTo(backup)
+                    backup
+                }
 
-                        buildFile?.let {
-                            buildFileBackup?.renameTo(it)
-                        }
+                gradle.afterMirakleEvaluate {
+                    settingsFile?.let {
+                        settingsBackup?.renameTo(it)
+                    }
+
+                    buildFile?.let {
+                        buildFileBackup?.renameTo(it)
+                    }
+
+                    versionCatalogBackup?.zip(versionCatalog)?.forEach { (backup, original) ->
+                        backup.renameTo(original)
                     }
                 }
             }

--- a/plugin/src/main/kotlin/utils.kt
+++ b/plugin/src/main/kotlin/utils.kt
@@ -1,4 +1,4 @@
-import groovy.lang.Closure
+
 import org.apache.tools.ant.taskdefs.condition.Os
 import org.gradle.BuildAdapter
 import org.gradle.BuildResult
@@ -125,3 +125,26 @@ fun findGradlewRoot(root: File): File? {
 fun loadProperties(file: File) = file.takeIf(File::exists)?.let {
     Properties().apply { load(it.inputStream()) }.toMap() as Map<String, String>
 } ?: emptyMap()
+
+fun Gradle.afterMirakleEvaluate(callback: () -> Unit) {
+    var wasCallback = false
+
+    gradle.rootProject {
+        it.afterEvaluate {
+            if (!wasCallback) {
+                wasCallback = true
+                callback()
+            }
+        }
+    }
+
+    // in case if build failed before project evaluation
+    addBuildListener(object : BuildAdapter() {
+        override fun buildFinished(p0: BuildResult) {
+            if (p0.failure != null && !wasCallback) {
+                wasCallback = true
+                callback()
+            }
+        }
+    })
+}

